### PR TITLE
CSP-783 GetCalendarEvents page and pagesize added

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Models/GetCalendarEventsModel.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using SFA.DAS.ApprenticeAan.Application.CalendarEvents.Queries.GetCalendarEvents;
+using SFA.DAS.ApprenticeAan.Application.Common;
+using SFA.DAS.ApprenticeAan.Application.Infrastructure.Configuration;
+
+namespace SFA.DAS.ApprenticeAan.Api.Models;
+
+public class GetCalendarEventsModel
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Models/GetCalendarEventsModel.cs:8]: Suggest rename to `GetCalendatEventsRequestModel` to differentiate between response and request models

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Models/GetCalendarEventsModel.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    [FromQuery]
+    public List<int> CalendarId { get; set; } = new List<int>();
+
+    [FromQuery]
+    public List<int> RegionId { get; set; } = new List<int>();
+
+    [FromQuery]
+    public int? Page { get; set; }
+
+    [FromQuery] public int? PageSize { get; set; }
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Models/GetCalendarEventsModel.cs:31]: New line after attribute

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/SFA.DAS.ApprenticeAan.Application.UnitTests.csproj [hunk 3]
```diff
@@ -25,6 +25,7 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\Shared\SFA.DAS.SharedOuterApi\SFA.DAS.SharedOuterApi.csproj" />
+    <ProjectReference Include="..\SFA.DAS.ApprenticeAan.Api\SFA.DAS.ApprenticeAan.Api.csproj" />
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/SFA.DAS.ApprenticeAan.Application.UnitTests.csproj:28]: This is not correct reference, is this needed?

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/CalendarEvents/Queries/GetCalendarEvents/GetCalendarEventsQueryTests.cs
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/CalendarEvents/Queries/GetCalendarEvents/GetCalendarEventsQueryTests.cs:1]: This should target the model and not the query, so the tests should be in the API project and not here. This will also allow you to remove the invalid project reference.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/QueryStringParameterBuilderTests.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+using SFA.DAS.ApprenticeAan.Application.Common;
+using SFA.DAS.ApprenticeAan.Application.Services;
+
+namespace SFA.DAS.ApprenticeAan.Application.UnitTests.Services;
+
+public class QueryStringParameterBuilderTests
+{
+
+    [Test, AutoData]
+    public void Operator_PopulatesModelFromParameters(Guid memberId, DateTime? fromDate, DateTime? toDate,
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/QueryStringParameterBuilderTests.cs:14]: Instead of one big test, we need a test per field with both cases where value is null and not null. This will improve your coverage as well.

**Review by DevenKShah — COMMENTED** [review]
The API project is wrongly reference in Application.UnitTests. I have added hints on how to fix this and also improve coverage

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+namespace SFA.DAS.ApprenticeAan.Api.UnitTests.Services;
+
+public class QueryStringParameterBuilderTests
+{
+
+    [Test, AutoData]
+    public void Operator_PopulatesModelFromParameters(Guid memberId, DateTime? fromDate, DateTime? toDate,
+        List<EventFormat> eventFormats, List<int> calendarIds, List<int> regionIds, int? page, int? pageSize)
+    {
+        var model = new GetCalendarEventsRequestModel
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:17]: Can you use `sut` for the target object name please

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs [hunk 6]
```diff
@@ -0,0 +1,227 @@
+﻿using AutoFixture.NUnit3;
+using FluentAssertions;
+using SFA.DAS.ApprenticeAan.Api.Models;
+using SFA.DAS.ApprenticeAan.Application.CalendarEvents.Queries.GetCalendarEvents;
+using SFA.DAS.ApprenticeAan.Application.Common;
+using SFA.DAS.ApprenticeAan.Application.Services;
+
+namespace SFA.DAS.ApprenticeAan.Api.UnitTests.Services;
+
+public class QueryStringParameterBuilderTests
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:10]: The target is in Api\Models folder, so this test file needs to be moved to similar folder in the unit test project and it needs to reflect the target name in the filename as well.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        query.FromDate.Should().BeNull();
+        query.ToDate.Should().BeNull();
+        query.Page.Should().BeNull();
+        query.PageSize.Should().BeNull();
+    }
+
+    [Test, AutoData]
+    public void Operator_PopulatesModelFromParametersDatesNull(Guid memberId)
+    {
+        var model = new GetCalendarEventsQuery
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs:61]: This test is not targeting the correct object. You have created instance of return type instead of source type.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+using FluentAssertions;
+using SFA.DAS.ApprenticeAan.Api.Models;
+using SFA.DAS.ApprenticeAan.Application.CalendarEvents.Queries.GetCalendarEvents;
+using SFA.DAS.ApprenticeAan.Application.Common;
+
+namespace SFA.DAS.ApprenticeAan.Api.UnitTests.Controllers.CalendarEventsControllerTests;
+public class GetCalendarEventsRequestModelTests
+{
+    [Test, AutoData]
+    public void Operator_PopulatesModelFromParameters(Guid memberId, DateTime? fromDate, DateTime? toDate, List<EventFormat> eventFormats, List<int> calendarIds, List<int> regionIds, int? page, int? pageSize)
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs:11]: Why are we testing operator in controller?

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+        if (calendarId != null)
+        {
+            model.CalendarId = new List<int> { calendarId.Value };
+        }
+
+        var parameters = QueryStringParameterBuilder.BuildQueryStringParameters(model);
+
+        parameters.TryGetValue("calendarId", out var calendarIdResult);
+        if (calendarIdResult != null)
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:147]: The calendarId is initialised to empty array so I am not sure when it would ever be null?

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+        if (regionId != null)
+        {
+            model.RegionId = new List<int> { regionId.Value };
+        }
+
+        var parameters = QueryStringParameterBuilder.BuildQueryStringParameters(model);
+
+        parameters.TryGetValue("regionId", out var regionIdResult);
+        if (regionIdResult != null)
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:174]: Same issue here as calendarId

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs (hunk 5 above)
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:17]: ok

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs (hunk 6 above)
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:10]: ok

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs (hunk 8 above)
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs:11]: I have it in the wrong folder, moving it.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs (hunk 9 above)
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:147]: But the parameter list won't have a match if the calendarId is a list of 0, so the calendarIdResult is null in that case.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs (hunk 10 above)
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Services/QueryStringParameterBuilderTests.cs:174]: But the parameter list won't have a match if the regionId is a list of 0, so the regionIdResult is null in that case.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs (hunk 7 above)
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs:61]: yes I named it completely wrong!

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Models/GetCalendarEventsQueryTests.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        query.ToDate.Should().Be(toDate?.ToString("yyyy-MM-dd"));
+        query.EventFormat.Should().BeEquivalentTo(eventFormats);
+        query.Page.Should().Be(page);
+        query.PageSize.Should().Be(pageSize);
+    }
+
+    [Test, AutoData]
+    public void Operator_PopulatesQueryFromModelWithNulls(Guid memberId)
+    {
+        var model = new GetCalendarEventsQuery
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Models/GetCalendarEventsQueryTests.cs:38]: I think this is still not correct as we are not really testing anything here. Ideally I would expect that you populate member id in the model and then do the conversion from model to query and then all your asserts would make sense.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Models/QueryStringParameterBuilderTests.cs
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Models/QueryStringParameterBuilderTests.cs:1]: This should be in the Service folder as the Static class is in the service folder.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs (hunk 7 above)
Comment (DevenKShah) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs]: I think this is still not correct as we are not really testing anything here. Ideally I would expect that you populate member id in the model and then do the conversion from model to query and then all your asserts would make sen

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Models/QueryStringParameterBuilderTests.cs
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Models/QueryStringParameterBuilderTests.cs:1]: ok fixed

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs (hunk 7 above)
Comment (MarkFCain) [reply src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/CalendarEventsControllerTests/GetCalendarEventsQueryTests.cs]: I've added a test that injects a autodata'd model, but not sure if that's what you meant.

**Review by DevenKShah — APPROVED** [review]

**Review by DevenKShah — APPROVED** [review]
